bpf: install proxy routes using Go, remove init.sh

ti-mo opened the pull request:
```
proxy: implement routing setup in Go

This commit ports the setup of routing table 2004 from init.sh to Go.

It does not remove the old logic from init.sh yet, as it is the last remaining
piece. Removing the script and its surrounding code will follow in a separate
commit.
```

```
init.sh: remove script

bye bye
```

Fixes: #26255

```release-note
bpf: install proxy routes using Go, remove init.sh
```

ti-mo committed:
proxy: implement routing setup in Go

This commit ports the setup of routing table 2004 from init.sh to Go.

It does not remove the old logic from init.sh yet, as it is the last remaining
piece. Removing the script and its surrounding code will follow in a separate
commit.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
init.sh: remove script

bye bye

Signed-off-by: Timo Beckers <timo@isovalent.com>

dylandreimerink commented on pkg/proxy/routes.go:
```diff
@@ -0,0 +1,102 @@
+var (
+	// Routing rule for traffic to proxy.
+	tproxyRule = route.Rule{
+		Priority: 9,
```
I don't know how to feel about this one. I know this is just taken from init.sh. But it is an oddly specific number.

ti-mo replied:
Would you like to see this assigned to a constant? Or a rationale behind the value 9?

dylandreimerink replied:
I would add a comment at least, if it makes sense inline there or else on a constant. That comment would ideally have a rationale, but if we can't find one it would be nice to know that as well. If its just arbitrary between 0 and 100 at the low side of things then someone doesn't have to be afraid to touch it in fear of some Indiana Jones trap springing at any moment.

ti-mo replied:
Oh, but that's half the fun. :wink: It's indeed as you described; we move the kernel's default pref 0 rule to pref 100 to clear up 0-99 for our own purposes. I've documented this in the code now.

markpash reviewed (approved):
bye bye init.sh! :champagne:

mhofstetter commented on pkg/proxy/proxy.go:
```diff
@@ -423,9 +423,35 @@ func (p *Proxy) SetProxyPort(name string, proxyType ProxyType, port uint16, ingr
+	return nil
+}
+
 // ReinstallRules is called by daemon reconfiguration to re-install proxy ports rules that
```
nit: adapt name in comment

ti-mo replied:
Thanks, good catch.

mhofstetter commented on pkg/proxy/routes.go:
```diff
@@ -0,0 +1,102 @@
+// packets to the proxy.
+func installRoutesIPv4() error {
+	if err := route.Upsert(route4); err != nil {
+		return fmt.Errorf("inserting ipv4 proxy route %v: %w", route4, err)
```
nit: change all error messages to something like `failed to insert ...`

ti-mo replied:
We usually log these as errors, and error wrapping makes 'failed' a little redundant. Anecdotally, we tend to avoid it in ebpf-go. For example, it comes out as `error="failed to insert ..."` or `x went wrong: failed to insert ...`.

IMO being terse is more readable since there's less reading to be done for unwinding the stack trace: `reticulating splines: shaving yaks: inserting ipv4 proxy route ...`, etc.

mhofstetter replied:
thanks for the explanation - seems reasonable to me!

mhofstetter reviewed (approved):
thanks @ti-mo :rocket: only two non-blocking nits from my side

tklauser reviewed (approved):
Nice work!

ti-mo committed:
datapath: simplify and test NodeEnsureLocalRoutingRule

I had trouble understanding the code, so I attempted to reduce its scope
and added some tests.

Generally, we only expect to find a rule pointing to table 255 with pref 0
(see init.sh move_local_rules() removed in a subsequent commit). This patch
directly issues a delete of a pref 0 rule instead of searching for it first,
reducing the chance of unintentionally removing other rules, making the logic
less fragile and easier to understand in the process.

Signed-off-by: Timo Beckers <timo@isovalent.com>
